File: trs-sdk-java/tests/ClientTest.java
package dev.trs.sdk;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class ClientTest {
    public static void main(String[] args) throws Exception {
        testHealthSubmitQuerySyncReplay();
        testValidationError();
        testTimeoutConnectionError();
        System.out.println("All trs-sdk-java tests passed.");
    }

    private static void testHealthSubmitQuerySyncReplay() throws Exception {
        TestNode node = new TestNode(0);
        try {
            TRSClient client = new TRSClient(node.baseUrl(), Duration.ofSeconds(2));
            HealthStatus health = client.health();
            assertEquals("ok", health.status(), "health.status");

            client.submit(record("g1", "Observation", Map.of("subject", "boot", "value", 1)));
            List<Map<String, Object>> rows = client.query(Map.of("type", "Observation"));
            assertEquals(1, rows.size(), "query size");

            var sync = client.sync(rows);
            assertEquals(1, sync.acceptedCount(), "sync accepted_count");

            Map<String, Object> replay = client.replay();
            assertTrue(replay.containsKey("coordination"), "replay has coordination");
        } finally {
            node.stop();
        }
    }

    private static void testValidationError() throws Exception {
        TestNode node = new TestNode(0);
        try {
            TRSClient client = new TRSClient(node.baseUrl(), Duration.ofSeconds(2));
            try {
                client.submit(record("bad", "Observation", Map.of()));
                throw new AssertionError("expected TRSValidationError");
            } catch (TRSValidationError expected) {
                assertTrue(true, "validation error observed");
            }
        } finally {
            node.stop();
        }
    }

    private static void testTimeoutConnectionError() {
        TestNode node = null;
        try {
            node = new TestNode(250);
            TRSClient client = new TRSClient(node.baseUrl(), Duration.ofMillis(30));
            try {
                client.health();
                throw new AssertionError("expected TRSConnectionError");
            } catch (TRSConnectionError expected) {
                assertTrue(true, "connection error observed");
            }
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        } finally {
            if (node != null) {
                node.stop();
            }
        }
    }

    private static Map<String, Object> record(String id, String type, Map<String, Object> payload) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("id", id);
        out.put("type", type);
        out.put("author", "tester");
        out.put("timestamp", "2026-08-04T10:00:00+00:00");
        out.put("schema", switch (type) {
            case "Observation" -> "trs.observation.v1";
            case "Intention" -> "trs.intention.v1";
            case "Commitment" -> "trs.commitment.v1";
            default -> "unknown";
        });
        out.put("payload", payload);
        out.put("causes", List.of());
        out.put("authorization", List.of());
        out.put("signature", "sig:" + id);
        return out;
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if ((expected == null && actual != null) || (expected != null && !expected.equals(actual))) {
            throw new AssertionError(label + " expected=" + expected + " actual=" + actual);
        }
    }

    private static void assertTrue(boolean value, String label) {
        if (!value) {
            throw new AssertionError(label);
        }
    }

    private static final class TestNode {
        private final HttpServer server;
        private final List<Map<String, Object>> records = new ArrayList<>();
        private final long delayMillis;

        private TestNode(long delayMillis) throws IOException {
            this.delayMillis = delayMillis;
            this.server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
            this.server.createContext("/health", this::handleHealth);
            this.server.createContext("/submit", this::handleSubmit);
            this.server.createContext("/query", this::handleQuery);
            this.server.createContext("/sync", this::handleSync);
            this.server.createContext("/replay", this::handleReplay);
            this.server.start();
        }

        private String baseUrl() {
            return "http://127.0.0.1:" + server.getAddress().getPort();
        }

        private void stop() {
            server.stop(0);
        }

        private void maybeDelay() {
            if (delayMillis <= 0) {
                return;
            }
            try {
                Thread.sleep(delayMillis);
            } catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
            }
        }

        private void handleHealth(HttpExchange exchange) throws IOException {
            maybeDelay();
            write(exchange, 200, "{\"status\":\"ok\",\"runtime\":\"1.0.0\",\"node\":\"0.1.0\"}");
        }

        @SuppressWarnings("unchecked")
        private void handleSubmit(HttpExchange exchange) throws IOException {
            maybeDelay();
            String body = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
            Object parsed = Jsons.parse(body);
            Map<String, Object> root = (Map<String, Object>) parsed;
            Map<String, Object> record = (Map<String, Object>) root.get("record");
            Map<String, Object> payload = (Map<String, Object>) record.get("payload");

            boolean valid = payload.containsKey("subject") || payload.containsKey("goal") || payload.containsKey("action");
            if (!valid) {
                write(
                        exchange,
                        200,
                        "{\"accepted\":false,\"record_id\":\""
                                + record.get("id")
                                + "\",\"errors\":[\"5.3 Payload Shape\"]}");
                return;
            }
            records.add(record);
            write(
                    exchange,
                    200,
                    "{\"accepted\":true,\"record_id\":\"" + record.get("id") + "\",\"errors\":[]}");
        }

        private void handleQuery(HttpExchange exchange) throws IOException {
            maybeDelay();
            String rows = Jsons.stringify(Map.of("records", records));
            write(exchange, 200, rows);
        }

        private void handleSync(HttpExchange exchange) throws IOException {
            maybeDelay();
            write(
                    exchange,
                    200,
                    "{\"accepted_count\":1,\"rejected_count\":0,\"appended_ids\":[\"g1\"],\"rejected_errors\":[]}");
        }

        private void handleReplay(HttpExchange exchange) throws IOException {
            maybeDelay();
            write(exchange, 200, "{\"coordination\":{\"unresolved_intentions\":[]}}");
        }

        private void write(HttpExchange exchange, int code, String json) throws IOException {
            byte[] payload = json.getBytes(StandardCharsets.UTF_8);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(code, payload.length);
            try (OutputStream out = exchange.getResponseBody()) {
                out.write(payload);
            }
        }
    }
}

File: trs-sdk-kotlin/tests/ClientTest.kt
package dev.trs.sdk

import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.time.Duration

fun main() {
    testHealthSubmitQuerySyncReplay()
    testValidationError()
    testTimeoutConnectionError()
    println("All trs-sdk-kotlin tests passed.")
}

private fun testHealthSubmitQuerySyncReplay() {
    val server = TestServer(delayMs = 0)
    try {
        val client = TrsClient(server.baseUrl, Duration.ofSeconds(2))
        val health = client.health()
        checkEquals("ok", health.status, "health.status")

        client.submit(
            mapOf(
                "id" to "g1",
                "type" to "Observation",
                "payload" to mapOf("subject" to "boot", "value" to 1),
            )
        )

        val rows = client.query(emptyMap())
        checkEquals(1, rows.size, "query size")
        val sync = client.sync(rows)
        checkEquals(1, sync.acceptedCount, "sync accepted_count")
        val replay = client.replay()
        checkTrue(replay.containsKey("coordination"), "replay contains coordination")
    } finally {
        server.close()
    }
}

private fun testValidationError() {
    val server = TestServer(delayMs = 0)
    try {
        val client = TrsClient(server.baseUrl, Duration.ofSeconds(2))
        try {
            client.submit(
                mapOf(
                    "id" to "bad",
                    "type" to "Observation",
                    "payload" to emptyMap<String, Any?>(),
                )
            )
            throw IllegalStateException("expected TrsValidationError")
        } catch (_: TrsValidationError) {
        }
    } finally {
        server.close()
    }
}

private fun testTimeoutConnectionError() {
    val server = TestServer(delayMs = 250)
    try {
        val client = TrsClient(server.baseUrl, Duration.ofMillis(20))
        try {
            client.health()
            throw IllegalStateException("expected TrsConnectionError")
        } catch (_: TrsConnectionError) {
        }
    } finally {
        server.close()
    }
}

private fun checkEquals(expected: Any, actual: Any, label: String) {
    if (expected != actual) {
        throw IllegalStateException("$label expected=$expected actual=$actual")
    }
}

private fun checkTrue(value: Boolean, label: String) {
    if (!value) {
        throw IllegalStateException("$label expected true")
    }
}

private class TestServer(private val delayMs: Long) {
    private val server: HttpServer = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
    private val records = mutableListOf<Map<String, Any?>>()

    val baseUrl: String

    init {
        server.createContext("/health") { exchange ->
            maybeDelay()
            respond(exchange, """{"status":"ok","runtime":"1.0.0","node":"0.1.0"}""")
        }
        server.createContext("/submit") { exchange ->
            maybeDelay()
            val body = exchange.requestBody.bufferedReader().use { it.readText().ifBlank { "{}" } }
            val root = Jsons.parse(body) as Map<*, *>
            val record = root["record"] as Map<*, *>
            val payload = record["payload"] as? Map<*, *> ?: emptyMap<String, Any?>()
            val valid = payload.containsKey("subject") || payload.containsKey("goal") || payload.containsKey("action")
            val id = record["id"]?.toString() ?: ""
            if (!valid) {
                respond(exchange, """{"accepted":false,"record_id":"$id","errors":["5.3 Payload Shape"]}""")
            } else {
                @Suppress("UNCHECKED_CAST")
                records.add(record as Map<String, Any?>)
                respond(exchange, """{"accepted":true,"record_id":"$id","errors":[]}""")
            }
        }
        server.createContext("/query") { exchange ->
            maybeDelay()
            respond(exchange, Jsons.stringify(mapOf("records" to records)))
        }
        server.createContext("/sync") { exchange ->
            maybeDelay()
            respond(exchange, """{"accepted_count":1,"rejected_count":0,"appended_ids":["g1"],"rejected_errors":[]}""")
        }
        server.createContext("/replay") { exchange ->
            maybeDelay()
            respond(exchange, """{"coordination":{"unresolved_intentions":[]}}""")
        }
        server.start()
        baseUrl = "http://127.0.0.1:${server.address.port}"
    }

    fun close() {
        server.stop(0)
    }

    private fun maybeDelay() {
        if (delayMs > 0) {
            Thread.sleep(delayMs)
        }
    }

    private fun respond(exchange: com.sun.net.httpserver.HttpExchange, body: String) {
        val payload = body.toByteArray(Charsets.UTF_8)
        exchange.responseHeaders.add("Content-Type", "application/json")
        exchange.sendResponseHeaders(200, payload.size.toLong())
        exchange.responseBody.use { out ->
            out.write(payload)
        }
    }
}

